File: ODE Solver/Python/expdecay.py
import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt

#define equations
def derivatives(t, y):
    #Get state variables
    v, y_var = y
    #Compute derivs
    dvdt = -g - k * v
    dydt = v 
    #update array with derivatives
    return [dvdt, dydt]

#Define parameters
g = 9.8 #gravity constant
k = 0.5 #drag coefficient

#Define initial conditions and timespan
time_span = [0,10]
y0 = [0 , 100] #initial values for v, y

#Call solver 
output = solve_ivp(derivatives, time_span, y0, method = 'BDF', dense_output = True)

#plot 
t_plot = np.linspace(0, 10, 100)
y_plot = output.sol(t_plot)
v_plot = y_plot[0]
y_var_plot = y_plot[1]
plt.figure()
plt.plot(t_plot, v_plot)
plt.figure()
plt.plot(t_plot, y_var_plot)
plt.show()

File: ODE Solver/Matlab/expdecay.m
%expdecay model with ode15s

%parameters
g = 9.8;
k = 0.5;
params = [g , k];
%initial conditions
t_span = [0 ; 10];
y0 = [0 ; 100]; % v, y
options = [];

%solvercall
[t, y] = ode15s(@derivatives, t_span, y0, options, params);

%plot 
plot(t, y(:,1));   %velocity over time

%equations
function output = derivatives(t, y, p)
    %unpack parameters
    g = p(1);
    k = p(2);
    %state variables
    v = y(1);

    %compute derivatives
    dvdt = -g - k * v;
    dydt = v;
    %return array of derivatives 
    output = [dvdt ; dydt];
end

